### PR TITLE
Configurable bulk action limits

### PR DESCRIPTION
**Pull request by ThoWagen**
This PR adds the option to disable or set limits for bulk actions.

It is based on the discussion in #3142

- [x] Fix errors in lightbox
- [x] Translate error messages
- [x] Find a solution for BulkActionTrait
- [x] Add Mink tests
- [x] refactor translations (wait for #3249)
- [x] Note BC breaks in changelog when merging (language file changes, CartController/MyResearchController constructor signature changes).
- Find and set proper limits for the actions (Task for a followup PR)

### CODE REVIEW

**Comment by ThoWagen** [comment]
@demiankatz This is a first version for this. There is still something to do (see todo list). I am not sure what the best solution is for the following problem.

Currently by using `redirectToSource` when the action limit is exceeded we may return to the result or favorite list inside of the lightbox. 

Some ideas on how we could fix this:
1. Additionally catching an exceeded action limit by using JS like we do when no items are selected. Advantage: We don't have to fist make a request and it's probably the user friendliest solution. Disadvantage: It might make the code less maintainable and more error prone.
2. We could create a new action that only renders the flashmessages. This could for example be added to the index controller. If we are in a lightbox we can set `$target` to this action `redirectToSource`. Advantage: It could also be used for similar problems. Disadvantage: It feels a bit hacky.
3. Catching the exceeded limit after the forms inside of the lightbox is submitted and just returning the form with the error message. Advantage: Probably the simplest solution. Disadvantage: It is not very user friendly since error is displayed rather late.

What do you think. Do you have a better idea?

**Comment by demiankatz** [comment]
Thanks for the progress on this, @ThoWagen -- I'll look at this more closely when I'm back in the office. In the meantime, I agree that the most user-friendly and efficient solution would be to check in the Javascript. Maybe we can do this generically with a `data-max-selected-items` element on the form (or something similar). As you say, we'll need to be careful to implement checks in both the front end and the back end to avoid inconsistencies and gaps, and maybe we should figure out a way to bypass the Javascript check in tests so we can add test coverage to the backend logic. If you'd like to start exploring this route, I certainly wouldn't object! Though if you prefer to do the option 3 "simplest approach" first, that's fine too -- because we can always do it that way initially and add the extra JS checks as a later enhancement. Option 2 is interesting, but I agree that it feels a bit hacky.

File: config/vufind/config.ini
Comment (demiankatz) [inline config/vufind/config.ini]: Typo: "you setup" --> "your setup"

File: config/vufind/config.ini
Comment (demiankatz) [inline config/vufind/config.ini]: There might be practical reasons to leave this here, but did you consider putting these settings in Export.ini instead? It seems like it might be more desirable to put these configurations in a more specific place, to reduce noise in config.ini and to remind people to set this value when adding new export methods.

File: module/VuFind/src/VuFind/Controller/CartController.php
Comment (demiankatz) [inline module/VuFind/src/VuFind/Controller/CartController.php]: Maybe a `checkBulkActionLimit($action)` support method would make this code less repetitive.

I'm also a little concerned about defaulting this to 0, since users who fail to update your config.ini when upgrading will suddenly lose all bulk options. Maybe we should add a separate config setting for "default bulk limit" and default the default to 100. What do you think?

File: module/VuFind/src/VuFind/Controller/CartController.php
Comment (demiankatz) [inline module/VuFind/src/VuFind/Controller/CartController.php]: Might be better to pass the Config Plugin Manager here so we can access multiple configurations (i.e. if you want to take my suggestion about moving settings to export.ini).

File: themes/bootstrap3/templates/myresearch/bulk-action-buttons.phtml
Comment (demiankatz) [inline themes/bootstrap3/templates/myresearch/bulk-action-buttons.phtml]: If we add more smarts to limits as suggested above, we'll probably need more view helper logic somewhere to apply equivalent behavior here. If you want to go that route, I can give it a little more thought and propose a location, if you don't already have ideas.

**Review by demiankatz — CHANGES_REQUESTED** [review]
Thanks, @ThoWagen -- I was able to give this a deeper review and had a few more thoughts...

File: config/vufind/config.ini
Comment (ThoWagen) [reply config/vufind/config.ini]: fixed

File: config/vufind/config.ini
Comment (ThoWagen) [reply config/vufind/config.ini]: Moved it. But just realized that the explanation is missing in the comment of export.ini. I will add it.

File: module/VuFind/src/VuFind/Controller/CartController.php
Comment (ThoWagen) [reply module/VuFind/src/VuFind/Controller/CartController.php]: > Maybe a checkBulkActionLimit($action) support method would make this code less repetitive.

Did this and added it to a new BulkActionTrait.

>I'm also a little concerned about defaulting this to 0, since users who fail to update your config.ini when upgrading will suddenly lose all bulk options. Maybe we should add a separate config setting for "default bulk limit" and default the default to 100.

Good Idea. Did this.

File: module/VuFind/src/VuFind/Controller/CartController.php
Comment (ThoWagen) [reply module/VuFind/src/VuFind/Controller/CartController.php]: Done

File: themes/bootstrap3/templates/myresearch/bulk-action-buttons.phtml
Comment (ThoWagen) [reply themes/bootstrap3/templates/myresearch/bulk-action-buttons.phtml]: I added a view helper for rendering bulk action buttons that checks the limits.

**Comment by ThoWagen** [comment]
@demiankatz I implemented your suggestions.
There is a new BulkAction trait that is used by the CartController, MyResearchController and the new BulkAction view helper. But there is one problem with it. 

I also moved the `redirectToSource` method inside. But it is not used by the BulkAction view helper. To pass the tests I added an empty `redirectToSource` method to it. But this is not a good design choice I'm afraid. Maybe I just have to remove the method from the trait again with the drawback that the `deleteAction` inside of the MyResearchController has to implement this again. Do you have a better idea for this?

I also think the templates could be simplified by creating a mapping of the actions to the icons and translation key inside of the view helper. And I am not sure if we need all of the different translations for the titles. The email button for example has the title "email_selected" in the favorites, "bookbag_email_selected" in the search results and "bookbag_email". What do you think?

File: config/vufind/config.ini
Comment (demiankatz) [inline config/vufind/config.ini:1864]: It might be worth adding a comment to highlight the fact that this default value will be used for any action lacking a more specific setting.

File: module/VuFind/src/VuFind/Controller/MyResearchControllerFactory.php
Comment (demiankatz) [inline module/VuFind/src/VuFind/Controller/MyResearchControllerFactory.php]: You can simplify this last line to `return parent::__invoke($container, $requestedName, [$configLoader]);` -- the AbstractBaseFactory is designed to accept additional constructor parameters through the `$options` array, which then allows us to reduce redundancy in its children.

File: module/VuFind/src/VuFind/Controller/CartControllerFactory.php
Comment (demiankatz) [inline module/VuFind/src/VuFind/Controller/CartControllerFactory.php]: The same factory simplification can be applied here (and I see that somebody, possibly me, neglected to do that in the first place :-) ).

File: module/VuFind/src/VuFind/Controller/Feature/BulkActionTrait.php
Comment (demiankatz) [inline module/VuFind/src/VuFind/Controller/Feature/BulkActionTrait.php]: Might be worth doing `$config = $this->configLoader->get('config');` so we don't have to call the configLoader multiple times when checking the default... though in practice it's a very fast lookup and shouldn't make much difference. I'll leave it up to your discretion. :-)

File: module/VuFind/src/VuFind/Controller/Feature/BulkActionTrait.php
Comment (demiankatz) [inline module/VuFind/src/VuFind/Controller/Feature/BulkActionTrait.php]: This trait depends on having a specific session container set up in `$this->session`. We should document that, and possibly also add a check here that throws an exception if it's missing. I realize that this particular method is only going to be used in the context of a controller, but since this trait is shared with a view helper, it seems like a potential source of problems or confusion. Should we consider splitting to two traits or moving some functionality back into the controller?

File: module/VuFind/src/VuFind/View/Helper/Bootstrap3/BulkAction.php
Comment (demiankatz) [inline module/VuFind/src/VuFind/View/Helper/Bootstrap3/BulkAction.php]: You could reduce this array to `compact('action', 'icon', 'content', 'limit', 'attributes')`.

File: module/VuFind/src/VuFind/View/Helper/Bootstrap3/BulkAction.php
Comment (demiankatz) [inline module/VuFind/src/VuFind/View/Helper/Bootstrap3/BulkAction.php]: I guess this was how you decided to address my concerns about redirectToSource -- but I think I would prefer to handle this on the trait side with an `isset($this->session)` type of check instead of relying on the client to stub out problematic methods. But maybe I misunderstand your intent here or am overlooking some important detail!

File: themes/bootstrap3/js/common.js
Comment (demiankatz) [inline themes/bootstrap3/js/common.js]: Maybe worth changing to `document.querySelector` to reduce jQuery usage while we're changing the line anyway.

File: themes/bootstrap3/templates/search/bulk-action-buttons.phtml
Comment (demiankatz) [inline themes/bootstrap3/templates/search/bulk-action-buttons.phtml]: $actionLimits is still being defined without being used in several templates... and it's used here, but I don't think it needs to be because the view helper now incorporates the limit check.

**Review by demiankatz — CHANGES_REQUESTED** [review]
Thanks, @ThoWagen, this is shaping up very nicely! I haven't done any hands-on testing yet, but the below comments (and a couple of recent trivial commits) are the result of a fairly thorough read-through. Once these are addressed, I'll work with @sturkel89 to actually exercise all the code!

**Comment by demiankatz** [comment]
@ThoWagen, my apologies -- I should have more carefully read your comments before writing my review, as I see you already raised the issue of the BulkActionTrait. One option might be to have a BulkActionControllerTrait which incorporates BulkActionTrait and adds the controller-specific methods. There might be some other repetitive logic which could be moved into such a trait (e.g. something like a `getSelectedIdsFromRequest()` method to encapsulate the repeated logic that picks the right posted ID values.

I also agree that moving more of the button settings into the view helper might make sense, though we should be careful that we don't make anything too hard to customize or override (e.g. if a user wants to add a local custom button, it shouldn't be too difficult to do). It does appear that we can clean up some of the redundant translation strings as well. Maybe that task should wait until we finish #3249, since doing so would enable us to set up aliases for some of the redundant strings -- then perhaps the view helper could include a "context-prefix" argument that would be applied to translation strings, and this would allow users to differentiate labels for bulk vs. cart if they wanted to, without creating a heavier burden in language file maintenance.

File: config/vufind/config.ini
Comment (ThoWagen) [reply config/vufind/config.ini:1864]: done

File: module/VuFind/src/VuFind/Controller/MyResearchControllerFactory.php
Comment (ThoWagen) [reply module/VuFind/src/VuFind/Controller/MyResearchControllerFactory.php]: done

File: module/VuFind/src/VuFind/Controller/CartControllerFactory.php
Comment (ThoWagen) [reply module/VuFind/src/VuFind/Controller/CartControllerFactory.php]: done

File: module/VuFind/src/VuFind/Controller/Feature/BulkActionTrait.php
Comment (ThoWagen) [reply module/VuFind/src/VuFind/Controller/Feature/BulkActionTrait.php]: done

File: module/VuFind/src/VuFind/Controller/Feature/BulkActionTrait.php
Comment (ThoWagen) [reply module/VuFind/src/VuFind/Controller/Feature/BulkActionTrait.php]: I split the trait into two as you suggested and added the dependency in the doc comment.

File: module/VuFind/src/VuFind/View/Helper/Bootstrap3/BulkAction.php
Comment (ThoWagen) [reply module/VuFind/src/VuFind/View/Helper/Bootstrap3/BulkAction.php]: done

File: module/VuFind/src/VuFind/View/Helper/Bootstrap3/BulkAction.php
Comment (ThoWagen) [reply module/VuFind/src/VuFind/View/Helper/Bootstrap3/BulkAction.php]: removed it, since it is not necessary anymore.

File: themes/bootstrap3/js/common.js
Comment (ThoWagen) [reply themes/bootstrap3/js/common.js]: done

File: themes/bootstrap3/templates/search/bulk-action-buttons.phtml
Comment (ThoWagen) [reply themes/bootstrap3/templates/search/bulk-action-buttons.phtml]: cleaned up the templates.

**Comment by ThoWagen** [comment]
I split the trait into two traits and moved the "main" trait into a new "Feature" directory in the VuFind module. But maybe there is a better place for it. 

Now, refactoring the translations and figuring out proper limit is still todo. For the translations I wait as suggested for #3249. The limits should be tested on multiple systems. Maybe it's better to stick with the limits of 100 for now and do it as a followup of this and #3142.

And I agree that it is better not to move to much logic into the view helper to keep adjustments simpler.

**Comment by demiankatz** [comment]
Thanks, @ThoWagen, the changes so far look good! I'll try to make some more progress on #3249 later this week (hopefully on Friday) and we can resume work here when that is done. I agree that it probably makes sense to start with default limits of 100, since that essentially mirrors the existing functionality, and we can investigate tweaking defaults as a separate project after the basic functionality is in place.

**Comment by demiankatz** [comment]
@ThoWagen, as noted elsewhere, #3249 is now merged, so you should be free to make further progress here. Please let me know if I can do more to help and/or when you're ready for me to do another round of tests and reviewing.

**Comment by ThoWagen** [comment]
@demiankatz I refactored the translations using aliases. All tasks should be completed now.

File: languages/aliases.ini
Comment (demiankatz) [inline languages/aliases.ini]: Would it be better to alias this directly to delete_selected instead of aliasing an alias? Or do you prefer it this way?

File: config/vufind/export.ini
Comment (demiankatz) [inline config/vufind/export.ini:17]: It's not exactly accurate to say that this limit overrides the [BulkActions] limit, since it only overrides if it is LOWER. If I set a limit of 5 in config.ini, exports will be blocked even if everything is set to 100 in this file. Not sure if we need to clarify this comment or change some behavior; I don't have strong feelings and am happy to use whatever approach is easiest, at least to begin with!

**Review by demiankatz — CHANGES_REQUESTED** [review]
Thanks, @ThoWagen! This looks good to me and all tests are passing. Just a couple of small things to discuss before it is merged.

File: languages/aliases.ini
Comment (ThoWagen) [reply languages/aliases.ini]: As long as we only have a bulk delete button in the favorites we can do this. Done.

File: config/vufind/export.ini
Comment (ThoWagen) [reply config/vufind/export.ini:17]: The export actually isn't blocked but you get a message that it is. To fix this we could either use the maximum of all export limits as the limit that is used for the message in the general case or we can implement that the general export limit in config.ini is an upper limit of the limit in export.ini and update this comment. I'm not sure what the better approach is.

Comment (demiankatz) [reply config/vufind/export.ini:17]: I could be persuaded to pick either of the approaches you propose.

I think having the export settings truly override the config.ini settings, whether they go higher or lower, seems like the most intuitive and flexible approach... but the disadvantage is that I suspect calculating the correct limit value may become complicated and require special-case logic somewhere in the code. If this proves too burdensome, I think the compromise of treating the config.ini setting as a hard maximum is acceptable.

I also think that regardless of which approach we take, it may be a good idea to comment out more of the export configurations by default. Right now, we have lots of settings all set to the same value of 100. If somebody wants to change it, they'll potentially need to override a lot of places. But if we just set the default to 100 and comment everything else out, we'll get the same effect, and it will be easier for people to customize. (Obviously, we might decide to uncomment more of these settings in future once we figure out appropriate individual defaults for each situation... but as a starting point, I think just having a default set will make things easier).

Comment (ThoWagen) [reply config/vufind/export.ini:17]: I implement the first approach and commented the limits in export.ini out.

File: module/VuFind/src/VuFind/View/Helper/Bootstrap3/BulkActionFactory.php
Comment (demiankatz) [inline module/VuFind/src/VuFind/View/Helper/Bootstrap3/BulkActionFactory.php]: Might be worth updating this copyright line.

File: module/VuFind/src/VuFind/View/Helper/Bootstrap3/BulkAction.php
Comment (demiankatz) [inline module/VuFind/src/VuFind/View/Helper/Bootstrap3/BulkAction.php]: If the only reason for adding the serviceLocator is so it can be used in `getBulkActionExport()`, maybe it would be better to directly inject and set `$this->bulkActionExport` instead. Exposing the service locator is generally a bad practice, and I'm trying to reduce that pattern as much as possible -- if we can avoid introducing it here, I think that would be preferable.

File: module/VuFind/src/VuFind/View/Helper/Bootstrap3/BulkAction.php
Comment (demiankatz) [inline module/VuFind/src/VuFind/View/Helper/Bootstrap3/BulkAction.php:30]: Is there a strong reason to put this in the Bootstrap3 namespace rather than the Root namespace? I realize that there might be theme-specific class names involved, but even if that's the case, it might make sense to have a Root version that gets extended/overridden in specific theme namespaces. I don't insist on changing this if you think it's best as-is, but since we're currently working on migrating from bootstrap3 to bootstrap5, I think it might reduce future work if we can keep things in the Root namespace.

**Review by demiankatz — CHANGES_REQUESTED** [review]
Thanks, @ThoWagen. This seems to be working as expected, and all tests are still passing.

See below for a couple of final thoughts related to the view helper; I think we can merge once we've sorted these out.

File: module/VuFind/src/VuFind/View/Helper/Bootstrap3/BulkActionFactory.php
Comment (ThoWagen) [reply module/VuFind/src/VuFind/View/Helper/Bootstrap3/BulkActionFactory.php]: Done

File: module/VuFind/src/VuFind/View/Helper/Bootstrap3/BulkAction.php
Comment (ThoWagen) [reply module/VuFind/src/VuFind/View/Helper/Bootstrap3/BulkAction.php]: The BulkActionTrait does depend of `$export` instead now.

File: module/VuFind/src/VuFind/View/Helper/Bootstrap3/BulkAction.php
Comment (ThoWagen) [reply module/VuFind/src/VuFind/View/Helper/Bootstrap3/BulkAction.php:30]: Moved most of the logic to Root and only setting the Bootstrap3 button class here.

File: module/VuFind/src/VuFind/Controller/CartController.php
Comment (demiankatz) [inline module/VuFind/src/VuFind/Controller/CartController.php:73]: Would it make more sense to move this property definition into the trait that uses it, rather than duplicating it in multiple places?

File: module/VuFind/src/VuFind/Feature/BulkActionTrait.php
Comment (demiankatz) [inline module/VuFind/src/VuFind/Feature/BulkActionTrait.php:96]: Maybe worth throwing an exception if export is unset at this stage?

**Review by demiankatz — CHANGES_REQUESTED** [review]
Thanks, @ThoWagen -- a couple final suggestions on the latest refactoring!

File: module/VuFind/src/VuFind/Controller/CartController.php
Comment (ThoWagen) [reply module/VuFind/src/VuFind/Controller/CartController.php:73]: \VuFind\Export is actually also used in CartController.php. I removed the getExport method since we set it in the constructor now.

I checked how other traits handle dependencies and those that I checked state the dependency in the doc comment but don't define the properties themselves. And since they get set in the constructors of the classes I think it is better to define them in the classes as well. But this more based on a feeling than on knowledge of trait design concepts. If you prefer to define `$export` in the trait I will do so.

File: module/VuFind/src/VuFind/Feature/BulkActionTrait.php
Comment (ThoWagen) [reply module/VuFind/src/VuFind/Feature/BulkActionTrait.php:96]: As with the comment above I checked other traits. They only state the dependency in the doc comment and do not check them. And in this case we get an exception anyway if we want to call the method on null. A custom exception might only have a better description. I will still add it if you want.

File: module/VuFind/src/VuFind/Controller/CartController.php
Comment (demiankatz) [reply module/VuFind/src/VuFind/Controller/CartController.php:73]: It's fine to leave it as-is. This is the part of traits that I simply do not like -- I wish that traits could formally define requirements and/or have some kind of constructor equivalent to initialize them... but the reality is that traits are a very simplistic mechanism, and these types of things don't really have a place there. We just have to do our best to come up with useful conventions to avoid problems.

Perhaps there is an argument for moving the export property definition into the trait and also adding an `initializeExportTrait` method which could be called from the class constructors to provide the required value(s). But that feels like inventing a whole new mechanism that doesn't already have a precedent, and now is probably not the time to do it. :-)

File: module/VuFind/src/VuFind/Feature/BulkActionTrait.php
Comment (demiankatz) [reply module/VuFind/src/VuFind/Feature/BulkActionTrait.php:96]: Similarly, it's fine to leave this as-is. I think in my mind, there's a difference between properties that the trait expects the composing class to already have, vs. things that are used primarily by the trait. But in reality this distinction is not usually very clear, so it does not make sense to overcomplicate things based on my feelings here.

**Review by demiankatz — APPROVED** [review]
Thanks, @ThoWagen, I think this is in good shape now. I fixed a couple of small typos in comments and made some updates to a recently-added language file, and I think it's now ready to merge. When you have a chance, I think we're now ready to resume work on #3142. In the meantime, I hope you have a good weekend!

**Pull request merged**
